src/mlpack/methods/ann/layer/multihead_attention.hpp

```diff
@@ -0,0 +1,246 @@
+ *   title   = {Attention Is All You Need},
+ *   year    = {2017},
+ *   url     = {http://arxiv.org/abs/1706.03762v5}
+ * }
+ * @endcode
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_HPP
+#define MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <boost/ptr_container/ptr_vector.hpp>
+
+#include "../visitor/delta_visitor.hpp"
+#include "../visitor/output_parameter_visitor.hpp"
+#include "../visitor/reset_visitor.hpp"
+#include "../visitor/weight_size_visitor.hpp"
+
+#include "layer_types.hpp"
+#include "add_merge.hpp"
+#include "sequential.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * This class implements the Multihead Attention Layer, using a variety of
+ * possible layer implementations.
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class MultiheadAttention
+{
+ public:
+  /**
+   * Default constructor: this will not give a usable MultiheadAttention object,
+   * so be sure to set all the parameters before use.
+   */
+  MultiheadAttention();
+
+  /**
+   * Create the MultiheadAttention object using the specified modules.
+   *
+   * @param embedDim Total dimension of the model.
+   * @param numHeads Number of parallel attention heads.
+   * @param dropout The dropout rate for attention output weights.
+   * @param bias Add bias as module parameter.
+   * @param addBiasKV Add bias to the key and value sequences.
+   * @param addZeroAttn Add a new batch of zeros to the key and value sequences.
+   * @param kdim Dimensionality of key.
+   * @param vdim Dimensionality of value.
+   */
+
+  MultiheadAttention(const size_t embedDim,
+                     const size_t numHeads,
+                     const float dropout = 0.0,
+                     const bool addZeroAttn = false,
+                     const size_t kdim = NULL,
+                     const size_t vdim = NULL);
+
+  /**
+   * Reset the layer parameters.
+   */
+  void ResetParameters();
+
+  /**
+   * Ordinary feed forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward through f.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   */
+  template<typename MatType = arma::mat>
+  void Forward(const arma::field<MatType>& input, arma::field<MatType>& output);
+
```

Comment:
Can we use field object as input and output in forward function? As the forward function needs three matrices as input. How should I go with this?

Reply:
I am not sure that's a good idea since each layer takes the output of the previous layer as the input data.

Reply:
Right. I will better try to vectorize the inputs or concatenate them. I'll see what works.

Reply:
@lozhnikov: I've changed the inputs to matrix types. Probably one main thing remains in forward function is softmax layer (PR #2351)